cmd/admin-handlers.go

```diff
@@ -256,47 +265,83 @@ func (adminAPI adminAPIHandlers) ServerInfoHandler(w http.ResponseWriter, r *htt
 	}
 	storage := objLayer.StorageInfo()
 
-	// Build list of enabled ARNs queues
-	var arns []string
-	for queueArn := range globalEventNotifier.GetAllExternalTargets() {
-		arns = append(arns, queueArn)
-	}
-
-	// Fetch uptimes from all peers. This may fail to due to lack
-	// of read-quorum availability.
-	uptime, err := getPeerUptimes(globalAdminPeers)
-	if err != nil {
-		writeErrorResponse(w, toAPIErrorCode(err), r.URL)
-		errorIf(err, "Unable to get uptime from majority of servers.")
-		return
-	}
-
-	// Build server properties information
-	properties := ServerProperties{
-		Version:  Version,
-		CommitID: CommitID,
-		Region:   serverConfig.GetRegion(),
-		SQSARN:   arns,
-		Uptime:   uptime,
-	}
-
-	// Build network info
-	connStats := ServerConnStats{
-		TotalInputBytes:  globalConnStats.getTotalInputBytes(),
-		TotalOutputBytes: globalConnStats.getTotalOutputBytes(),
-	}
-	httpStats := globalHTTPStats.toServerHTTPStats()
-
-	// Build the whole returned information
-	info := ServerInfo{
-		StorageInfo: storage,
-		ConnStats:   connStats,
-		HTTPStats:   httpStats,
-		Properties:  properties,
-	}
+	// Web service response
+	reply := make([]ServerInfo, len(globalAdminPeers))
+
+	var wg sync.WaitGroup
+
+	// Gather server information for all nodes
+	for i, p := range globalAdminPeers {
+		wg.Add(1)
+
+		// Gather information from a peer in a goroutine
+		go func(idx int, peer adminPeer) {
+			defer wg.Done()
+
+			// Initialize server info at index
+			reply[idx] = ServerInfo{Addr: peer.addr}
+
+			uptime, err := peer.cmdRunner.Uptime()
+			if err != nil {
+				errorIf(err, "Unable to get uptime from %s.", peer.addr)
+				reply[idx].Error = err
+				return
+			}
+
+			version, commitID, err := peer.cmdRunner.Version()
+			if err != nil {
+				errorIf(err, "Unable to get version from %s.", peer.addr)
+				reply[idx].Error = err
+				return
+			}
+
+			region, err := peer.cmdRunner.Region()
+			if err != nil {
+				errorIf(err, "Unable to get region from %s.", peer.addr)
+				reply[idx].Error = err
+				return
+			}
+
+			arns, err := peer.cmdRunner.ARNs()
+			if err != nil {
+				errorIf(err, "Unable to get ARNs from %s", peer.addr)
+				reply[idx].Error = err
+				return
+			}
+
+			connStats, err := peer.cmdRunner.ConnStats()
+			if err != nil {
+				errorIf(err, "Unable to get connection stats from %s.", peer.addr)
+				reply[idx].Error = err
+				return
+			}
+
+			httpStats, err := peer.cmdRunner.HTTPStats()
+			if err != nil {
+				errorIf(err, "Unable to get http stats from %s.", peer.addr)
+				reply[idx].Error = err
+				return
+			}
```

Comment:
The problem is that we are doing independent calls, it can be a place where users can DDOS the server.  Since our server mux would create more conn instances we should avoid it. 

It is better to make a bulk call since all of these info are part of the same struct, we should only do them separately if they are truly independent like Heal or Service.